### PR TITLE
feature/se-8-create ACL element

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|-----:|-----:|-------:|
| ETA                  | 20.8   |    0 |     0 |      0 |        0 |
| Developing      |  7+8+2.5+1  |    0 |    0 |      0 |         0 |
| Review             |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -   |   -     |        0 |
## Problem description
The submission can contain admin only fields with fined-grained permissions.

## Problems
- I wanted to get the actual context by method _contextService.getContext()_ but it returns null. I found out the context could be obtained from the HttpRequest by method _ContextUtil.obtainCurrentRequestContext()_.

### CODE REVIEW

File: dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+    private String granteeID;
+
+    /**
+     * Creates new ACE object from given String
+     *
+     * @param s
+     * @return ACE object or null
+     */
+
```
Comment (vidiecan) [inline dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java:76]: is the newline according to our coding style?
does java(/dspace) has something like a codestyle linter?

File: dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Convenience method to verify if this entry is allowing the action;
+     *
+     * @return
```
Comment (vidiecan) [inline dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java:189]: ?

File: dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+import java.util.Set;
+
+import org.apache.log4j.Logger;
+
+/**
+ * Class that represents single Access Control Entry
+ *
+ * @author Michal Josífko
+ * @author milanmajchrak
```
Comment (vidiecan) [inline dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java:18]: update the file according to a basic code style

Comment (vidiecan) [reply dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java:18]: mention @ufal (LINDAT/CLARIN) clarin-dspace

File: dspace-api/src/main/java/sk/dtq/dspace/app/util/ACL.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+import org.dspace.eperson.Group;
+import org.dspace.eperson.factory.EPersonServiceFactory;
+import org.dspace.eperson.service.GroupService;
+import org.springframework.stereotype.Component;
+
+/**
+ * Class that represents Access Control List
+ *
+ * @author Michal Josífko
+ * @author milanmajchrak
```
Comment (vidiecan) [inline dspace-api/src/main/java/sk/dtq/dspace/app/util/ACL.java:32]:
```suggestion
 * @author Milan Majchrak ( @ dataquest.sk)
```

File: dspace-api/src/main/java/sk/dtq/dspace/app/util/ACL.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+import org.dspace.core.Context;
+import org.dspace.eperson.EPerson;
+import org.dspace.eperson.Group;
+import org.dspace.eperson.factory.EPersonServiceFactory;
+import org.dspace.eperson.service.GroupService;
+import org.springframework.stereotype.Component;
+
+/**
+ * Class that represents Access Control List
+ *
```
Comment (vidiecan) [inline dspace-api/src/main/java/sk/dtq/dspace/app/util/ACL.java:30]:
```suggestion
 * @author LINDAT/CLARIAH-CZ (https://github.com/ufal/clarin-dspace)
```

File: dspace-api/src/test/java/sk/dtq/dspace/app/util/ACLTest.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    @Override
+    public void init() {
+        super.init();
+
+        // Initialize our spy of the autowired (global) authorizeService bean.
+        // This allows us to customize the bean's method return values in tests below
+        authorizeServiceSpy = spy(authorizeService);
+        groupServiceSpy = spy(groupService);
+
+//        ReflectionTestUtils.setField(ePersonService, "authorizeService", authorizeServiceSpy);
```
Comment (vidiecan) [inline dspace-api/src/test/java/sk/dtq/dspace/app/util/ACLTest.java:61]: why // ?

File: dspace/config/submission-forms.xml [hunk 7]
```diff
@@ -137,6 +137,9 @@
                     <input-type>series</input-type>
                     <hint>Enter the series and number assigned to this item by your community.</hint>
                     <required></required>
+<!--                    <acl>-->
+<!--                        policy=deny,action=read,grantee-type=user,grantee-id=*-->
```
Comment (vidiecan) [inline dspace/config/submission-forms.xml:141]: why?

File: dspace/config/submission-forms.xml [hunk 8]
```diff
@@ -152,6 +155,9 @@
                         it, please enter the types and the actual numbers or codes.
                     </hint>
                     <required></required>
+<!--                    <acl>-->
+<!--                        policy=allow,action=read,grantee-type=user,grantee-id=*-->
```
Comment (vidiecan) [inline dspace/config/submission-forms.xml:159]: why?

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-api/src/main/java/sk/dtq/dspace/app/util/ACL.java (hunk 4 above)
Comment (milanmajchrak) [reply dspace-api/src/main/java/sk/dtq/dspace/app/util/ACL.java:32]: changed to @author Milan Majchrak (milan.majchrak at dataquest dot sk)

File: dspace-api/src/main/java/org/dspace/app/util/DCInput.java [hunk 9]
```diff
@@ -536,6 +541,20 @@ public List<String> getExternalSources() {
         return externalSources;
     }
 
+    /**
+     * Is user allowed for particular ACL action on this input field in given
+     * Context?
+     *
+     * @param c
```
[the remark this answers is not in the document]
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/app/util/DCInput.java:548]: changed to useful comment

File: dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java (hunk 2 above)
Comment (milanmajchrak) [reply dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java:189]: changed to useful comment

File: dspace-api/src/main/java/sk/dtq/dspace/app/util/ACL.java (hunk 5 above)
Comment (milanmajchrak) [reply dspace-api/src/main/java/sk/dtq/dspace/app/util/ACL.java:30]: Added source class information:
_Class is copied from the LINDAT/CLARIAH-CZ (https://github.com/ufal/clarin-dspace) and modified by
 @author Milan Majchrak (milan.majchrak at dataquest dot sk)_

File: dspace/config/submission-forms.xml (hunk 8 above)
Comment (milanmajchrak) [reply dspace/config/submission-forms.xml:159]: Removed

Comment (milanmajchrak) [reply dspace/config/submission-forms.xml:159]: Added example acl definition to the form definition file

File: dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package sk.dtq.dspace.app.util;
```
[the remark this answers is not in the document]
Comment (milanmajchrak) [reply dspace-api/src/main/java/sk/dtq/dspace/app/util/ACE.java:8]: dtq package removed and the classes was moved to the dspace packages

**Review by vidiecan — APPROVED** [review]

**Review by vidiecan — APPROVED** [review]

**Review by vidiecan — APPROVED** [review]

**Pull request merged**